### PR TITLE
feat: Image search example

### PR DESCRIPTION
**Pull request by Green-Sky**
I used the [usearch](https://github.com/unum-cloud/usearch) vector database library, to build a image embedding database and then to search in it by similarity using a text string.

build the database:
```
$ bin/image-search-build -m ../models/ggml-model-f16.bin ../tests/
clip_model_load: loading model from '../models/ggml-model-f16.bin' - please wait....................................................clip_model_load: model size =   288.93 MB / num tensors = 397
clip_model_load: model loadded
main: starting base dir scan of '../tests/'
main: found image file '../tests/white.jpg'
main: found image file '../tests/red_apple.jpg'
```
it generates an `images.usearch` and `images.path` file.

search by similarity to string:
```
$ bin/image-search -m ../models/ggml-model-f16.bin "a red apple"
clip_model_load: loading model from '../models/ggml-model-f16.bin' - please wait....................................................clip_model_load: model size =   288.93 MB / num tensors = 397
clip_model_load: model loadded
search results:
similarity path
  0.655244 /home/green/workspace/clip.cpp/tests/red_apple.jpg
  0.773273 /home/green/workspace/clip.cpp/tests/white.jpg
```

It is a very rough implementation. It requires `c++17` and `std::filesystem` for iterating the filesystem.

A similarity search via an image would be nice too.

basically we could copy all the features from here: https://github.com/yurijmikhalevich/rclip

### TODO:
 - [x] add README.md to example
 - [x] make it compile conditionally (bc c++17 and deps)
 - [ ] solve the usearch memory leak (tracking https://github.com/unum-cloud/usearch/issues/138)
 - [x] tabs to spaces
 - [x] use fetch content instead of committing deps

### CODE REVIEW

File: examples/image-search/usearch/index.hpp
Comment (monatis) [inline examples/image-search/usearch/index.hpp:1]: Instead of copying this file into the repo, you can configure the example's `CMakeLists.txt` to automatically download it. Citing from USearch:

> Alternatively fetch it with CMake:

```cmake

FetchContent_Declare(usearch GIT_REPOSITORY https://github.com/unum-cloud/usearch.git)
FetchContent_MakeAvailable(usearch)
```

And of course, git-ignore it to not accidentally push.

File: examples/image-search/build.cpp [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	const size_t vec_dim = clip_ctx->vision_model.hparams.projection_dim;
+
+	size_t label = 0;
+
+	// search for images in path and create embedding
+	for (const auto& base_dir : params.image_directories) {
+		fprintf(stdout, "%s: starting base dir scan of '%s'\n", __func__, base_dir.c_str());
+
+		for (auto const& dir_entry : std::filesystem::recursive_directory_iterator(base_dir)) {
+			if (!dir_entry.is_regular_file()) {
```
Comment (monatis) [inline examples/image-search/build.cpp:100]: Do we need C++17 just for this? According to their header, USearch requires only C++11. And implementing this recursive directory walk should not be that hard in C++11 as well.

File: examples/image-search/build.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+			clip_image_f32 img_res;
+			if (!clip_image_load_from_file(img_path, img0)) {
+				fprintf(stderr, "%s: failed to load image from '%s'\n", __func__, img_path.c_str());
+				continue;
+			}
+
+			clip_image_preprocess(&img0, &img_res);
+
+			std::vector<float> vec(vec_dim);
+			if (!clip_image_encode(clip_ctx, params.n_threads, img_res, vec.data())) {
```
Comment (monatis) [inline examples/image-search/build.cpp:126]: Why not using a float array directly? We are working with fixed lengths at the end.

**Review by monatis — CHANGES_REQUESTED** [review]
I like the idea to provide an image search example with tiny vector search implementation directly in C++, but I highlighted some points.

Also, let's not use tab characters for indents please.

>  general code clean up. but the rest of the repo is still in a similar state 😝

Things may be messy at early stages to quickly introduce new features, but I'm trying to tidy up the code and README as we make progress in the project. You can always suggest better structures for the code and the overall project.

**Comment by Green-Sky** [comment]
> Also, let's not use tab characters for indents please.

Ohhhhh yea. sorry. for got to reverse this :laughing:

File: examples/image-search/build.cpp (hunk 2 above)
Comment (Green-Sky) [reply examples/image-search/build.cpp:126]: are we? I thought it depends on the specific model used.
but yea, i guess we could reuse the same vector, since it does not change during execution.

File: examples/image-search/build.cpp (hunk 1 above)
Comment (Green-Sky) [reply examples/image-search/build.cpp:100]: c++17 filesystem was the easy route. with c++11 we have to use platform specific code. so different code paths for windows/linux etc.

File: examples/image-search/usearch/index.hpp
Comment (Green-Sky) [reply examples/image-search/usearch/index.hpp:1]: oh. that actually makes sense too. I forgot that this is at compile/configure time. when I looked at it my brain just assumed it meant we have to include the whole repo.
technically we only want that 1 file. and that usearch repo is kinda full of stuff, but will look into this. I think the cmake version needs to be increased for this too.

**Comment by Green-Sky** [comment]
> Things may be messy at early stages to quickly introduce new features, but I'm trying to tidy up the code and README as we make progress in the project. You can always suggest better structures for the code and the overall project.

yea, I also intentionally maked this PR as draft for those reasons. :)

File: examples/image-search/build.cpp (hunk 2 above)
Comment (monatis) [reply examples/image-search/build.cpp:126]: Yeah, it's more like a coding style issue. both this repo and the USearch repo use float arrays instead of `std::vector<float>` to represent embedding vectors.

File: examples/image-search/build.cpp (hunk 1 above)
Comment (monatis) [reply examples/image-search/build.cpp:100]: I'll look into this --if there's no gotcha other than different routes for Windows and unix, I can PR to this branch. I'd like just to support as many people as possible 😀

File: examples/image-search/usearch/index.hpp
Comment (monatis) [reply examples/image-search/usearch/index.hpp:1]: Good concern, but if my memory serves me right fetchContent was introduced in CMake 3.11, and we are setting the minimum required version to 3.12 already.

**Comment by monatis** [comment]
> yea, I also intentionally maked this PR as draft for those reasons. :)

Great, thank you. Your work really appreciated. And to be honest, I was not aware of Usearch, but instead considering the original hnswlib library for the search example. But Usearch looks like a better candidate.

File: examples/image-search/build.cpp (hunk 2 above)
Comment (Green-Sky) [reply examples/image-search/build.cpp:126]: > float arrays

but the size is not known at compile time. runtime size declaration is an extension, and not proper c++

Comment (monatis) [reply examples/image-search/build.cpp:126]: I just meant, for example, copying from `examples/main.cpp`

```C++
    float img_vec[vec_dim];
    if (!clip_image_encode(ctx, params.n_threads, img_res, img_vec))
    {
        return 1;
    }
```

But ok, it was just a nit, and not a strong opinion :).

**Comment by Green-Sky** [comment]
forcepushed with `FetchContent` and spaces instead of tabs.

File: examples/image-search/search.cpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    const size_t vec_dim = clip_ctx->vision_model.hparams.projection_dim;
+
+    auto tokens = clip_tokenize(clip_ctx, params.search_text);
+
+    std::vector<float> txt_vec(vec_dim);
+
+    clip_text_encode(clip_ctx, params.n_threads, tokens, txt_vec.data());
+
+    auto results = embd_index.search({txt_vec.data(), txt_vec.size()}, 5);
+
```
Comment (monatis) [inline examples/image-search/search.cpp:104]: Maybe also getting the number of results from CLI arg too?

Comment (Green-Sky) [reply examples/image-search/search.cpp:104]: yes. i could swear i left a TODO somewhere in the code...

File: examples/image-search/build.cpp (hunk 1 above)
Comment (Green-Sky) [reply examples/image-search/build.cpp]: I am not targeting merging WITH c++17, but the example disabled by default in cmake.
You(or me) can then patch it later to not use `std::filesystem`.

Comment (monatis) [reply examples/image-search/build.cpp]: Good idea! We can invest more in this later.

Feel free to mark it as ready for review once you're satisfied.

**Comment by monatis** [comment]
Do you need an extra hand in this?

**Comment by Green-Sky** [comment]
Was doing something completely different last week :)
Will try to get this into a mergeable state this week.

**Comment by Green-Sky** [comment]
just rebased and changed some stuff you committed on main

issues that I found:
1. you added cmake options for conditionally disabling tests and examples, i negated them and initialized them to standalone (more convenient)
2. you require c++20 (!) should only be 11 i think. (and 17 for fs in image-search until its changed to something else)
3. you set gcc/clang specific warning and `O3` optimization level indiscriminately https://github.com/monatis/clip.cpp/blob/1f140e027a3b9165a1edcdd4948bb44c3c2eacff/CMakeLists.txt#L136 (you should not, you default to Release https://github.com/monatis/clip.cpp/blob/1f140e027a3b9165a1edcdd4948bb44c3c2eacff/CMakeLists.txt#L7 , which sets `O3`)

**Comment by monatis** [comment]
Yes `CMakeLists.txt` still requires some testing and optimizations (also for other arches like different arm versions) so it might be another PR when I take down my Raspberry PIs from the shelf. C++20 standard was just a guard but we can downgrade it now or after some tests. I can refactor the directory walk with the function from `common-clip` that does not require C++17. Then everything will be ok with C++11.

I'm ok with it as is, and then we can continue to further improve it. When #30 is merged, we can also the use batch inference when indexing images.

**Comment by Green-Sky** [comment]
> Yes `CMakeLists.txt` still requires some testing and optimizations (also for other arches like different arm versions) so it might be another PR when I take down my Raspberry PIs from the shelf. C++20 standard was just a guard but we can downgrade it now or after some tests. I can refactor the directory walk with the function from `common-clip` that does not require C++17. Then everything will be ok with C++11.

yea, there have also been changes upstream (especially llama) , where you obviously ~copied~ based the code on.

> I'm ok with it as is, and then we can continue to further improve it. When #30 is merged, we can also the use batch inference when indexing images.

Yea. Let me run some small tests and then i will make it ready for review.

**Comment by monatis** [comment]
> yea, there have also been changes upstream (especially llama) , where you obviously copied based the code on.

Exactly. The entire ggml-based projects are moving very fast, so I believe we don't need to target 100% production-grade multiplatform-ability. The speed of iterations and innovations is prioritized currently.

> Yea. Let me run some small tests and then i will make it ready for review.

Great. Thank you.

**Comment by monatis** [comment]
USearch reports distance values instead of similarities, so I made changes to emphasize it. LGTM, merging. TODOs can be fixed later.

Thanks for your contributions.
